lib/galaxy/model/data_tables/mapping.py

```diff
@@ -0,0 +1,117 @@
+"""
+Details of how the data model objects are mapped onto the relational database
+are encapsulated here.
+"""
+
+import logging
+
+from sqlalchemy import (
+    Column,
+    DateTime,
+    ForeignKey,
+    Integer,
+    MetaData,
+    String, Table,
+    UniqueConstraint
+)
+from sqlalchemy.orm import mapper, relation
+
+from galaxy.model import data_tables as data_tables_model
+from galaxy.model.base import ModelMapping
+from galaxy.model.orm.engine_factory import build_engine
+from galaxy.model.orm.now import now
+
+log = logging.getLogger(__name__)
+
+metadata = MetaData()
+
+# Data Table in database tables
+data_tables_model.DataTable.table = Table(
+    'data_table',
+    metadata,
+    Column("id", Integer, primary_key=True),
+    Column("name", String(255), index=True),
+    Column("create_time", DateTime, default=now),
+    Column("update_time", DateTime, default=now, onupdate=now),
+    UniqueConstraint('name'),
+)
+
+data_tables_model.DataTableColumn.table = Table(
+    'data_table_column',
+    metadata,
+    Column("id", Integer, primary_key=True),
+    Column("name", String(255), index=True),
+    Column("create_time", DateTime, default=now),
+    Column("update_time", DateTime, default=now, onupdate=now),
+    UniqueConstraint('name'),
+)
+
+data_tables_model.DataTableColumnAssociation.table = Table(
+    'data_table_column_association',
+    metadata,
+    Column("id", Integer, primary_key=True),
+    Column("data_table_id", Integer, ForeignKey("data_table.id"), index=True),
+    Column("data_table_column_id", Integer, ForeignKey("data_table_column.id"), index=True),
+)
+
+data_tables_model.DataTableRow.table = Table(
+    'data_table_row',
+    metadata,
+    Column("id", Integer, primary_key=True),
+    Column("create_time", DateTime, default=now),
+    Column("update_time", DateTime, default=now, onupdate=now),
+)
+
+data_tables_model.DataTableRowAssociation.table = Table(
+    'data_table_row_association',
+    metadata,
+    Column("id", Integer, primary_key=True),
+    Column("data_table_id", Integer, ForeignKey("data_table.id"), index=True),
+    Column("data_table_row_id", Integer, ForeignKey("data_table_row.id"), index=True),
+)
+
+data_tables_model.DataTableField.table = Table(
```

Comment:
I don't think I agree with the idea of a generic table. This means there are #col * # rows rows in this table and doesn't allow leveraging row level optimizations when querying. How hard would it be to dynamically generate the table schema based on the description in the tool_data_table_conf.xml?

Reply:
SQL databases are optimized to work in rows. Your comment on not leveraging the row level optimization is actually incorrect. I leverage those. This dynamically scalable approach can be used with some more advanced SQL queries to create a single table with all the values. The function `get_rows_by_values` (https://github.com/galaxyproject/galaxy/blob/af7697c9ca7839fa1e6b4a5d05c7c5ebcc3a91ba/lib/galaxy/tools/data/__init__.py#L1183) creates a temporary table to handle these types of queries.    
   
In the end this is more a question of where do you add most difficulty, in the queries or in the initial setup. I chose to get more difficult queries as that's easier to handle than a more difficult initial setup. The current setup doesn't harm scalability nor does it hamper possible queries that you may want to execute. It merely adds a few "WITH" queries in order to view the table in an easier format.

Reply:
How does your generic table handle scanning a data table with a 1,000,000 rows?

Reply:
At ~200000 entries within SQLite it takes an average of 0.19 seconds to retrieve a row based on all entries within the same row.